modules/objdetect/src/qrcode.cpp

```diff
@@ -2486,17 +2486,24 @@ cv::String QRCodeDetector::decode(InputArray in, InputArray points,
     CV_Assert(src_points.size() == 4);
     CV_CheckGT(contourArea(src_points), 0.0, "Invalid QR code source points");
 
+    size_t straight_qrcode_channels = straight_qrcode.fixedType() ?
+                                      straight_qrcode.channels() : 1;
+    CV_Assert(straight_qrcode_channels == 1);
+
     QRDecode qrdec;
     qrdec.init(inarr, src_points);
     bool ok = qrdec.straightDecodingProcess();
 
     std::string decoded_info = qrdec.getDecodeInformation();
-
-    if (ok && straight_qrcode.needed())
+    if(!ok && straight_qrcode.needed())
+    {
+        straight_qrcode.release();
+    }
+    else if (straight_qrcode.needed())
     {
         qrdec.getStraightBarcode().convertTo(straight_qrcode,
                                              straight_qrcode.fixedType() ?
-                                             straight_qrcode.type() : CV_32FC2);
+                                             straight_qrcode.type() : CV_8UC1);
```

Comment:
Perhaps we should ignore this `.fixedType()` check and force `8UC1` only.
Exception would be raised in case of type mismatch (by `convertTo`).

The assertion check above about channels is not necessary too.

Reply:
@alalek `convertTo` isn't raise a exception in case type or channel number mismatch, but there will be exception later `Mat::type() == traits::Type<_Tp>::value` if we don't use `.fixedType()` check.
By this reason, it is better to continue use assertion check about channels number and `.fixedType()` check